File: solution/java/day07/src/main/java/ci/Pipeline.java
package ci;

import ci.dependencies.Config;
import ci.dependencies.Emailer;
import ci.dependencies.Logger;
import ci.dependencies.Project;

public class Pipeline {
    private static final String SUCCESS = "success";
    private final Config config;
    private final Emailer emailer;
    private final Logger log;

    public Pipeline(Config config, Emailer emailer, Logger log) {
        this.config = config;
        this.emailer = emailer;
        this.log = log;
    }

    public void run(Project project) {
        if (haveTestsFailed(project)) {
            sendEmail("Tests failed");
            return;
        }
        if (hasDeploymentFailed(project)) {
            sendEmail("Deployment failed");
            return;
        }
        sendEmail("Deployment completed successfully");
    }

    private boolean haveTestsFailed(Project project) {
        if (!project.hasTests()) {
            log.info("No tests");
            return false;
        }
        if (!SUCCESS.equals(project.runTests())) {
            log.error("Tests failed");
            return true;
        }
        log.info("Tests passed");
        return false;
    }

    private boolean hasDeploymentFailed(Project project) {
        if (SUCCESS.equals(project.deploy())) {
            log.info("Deployment successful");
            return false;
        }
        log.error("Deployment failed");
        return true;
    }

    private void sendEmail(String text) {
        if (!config.sendEmailSummary()) {
            log.info("Email disabled");
            return;
        }
        log.info("Sending email");
        emailer.send(text);
    }
}

File: solution/kotlin/day07/src/main/kotlin/ci/Pipeline.kt
package ci

import arrow.core.Option
import arrow.core.Some
import ci.TestStatus.NoTests
import ci.TestStatus.PassingTests

class Pipeline(
    private val config: Config,
    private val emailSender: Emailer,
    private val log: Logger
) {
    private val success = "success"

    fun run(project: Project) {
        createProjectContext(project)
            .flatMap { runTests(it) }
            .flatMap { deploy(it) }
    }

    private fun createProjectContext(project: Project): Option<PipelineContext> =
        Some(PipelineContext(project))
            .onSome { context -> if (!context.hasTests) log.info("No tests") }

    private fun runTests(context: PipelineContext): Option<PipelineContext> =
        if (context.hasTests) runTestsSafely(context)
        else Some(context)

    private fun runTestsSafely(context: PipelineContext): Option<PipelineContext> =
        Some(context.project.runTests())
            .map { testsResult -> context.withTestsResults(testsResult == success) }
            .onSome { logTestResult(it) }

    private fun logTestResult(context: PipelineContext) {
        if (!context.testsRanSuccessfully) {
            log.error("Tests failed")
            sendEmail("Tests failed")
            return
        }
        log.info("Tests passed")
    }

    private fun deploy(context: PipelineContext): Option<PipelineContext> =
        if (context.mustRunDeployment()) deploySafely(context) else Some(context)

    private fun deploySafely(context: PipelineContext): Option<PipelineContext> =
        Some(context.project.deploy())
            .map { deploymentResult -> context.withDeploymentStatus(deploymentResult == success) }
            .onSome { logDeploymentResult(it) }

    private fun logDeploymentResult(context: PipelineContext) {
        if (!context.deployedSuccessfully) {
            log.error("Deployment failed")
            sendEmail("Deployment failed")
            return
        }
        log.info("Deployment successful")
        sendEmail("Deployment completed successfully")
    }

    private fun sendEmail(text: String) {
        if (!config.sendEmailSummary()) {
            log.info("Email disabled")
            return
        }
        log.info("Sending email");
        emailSender(text);
    }
}

fun interface Config {
    fun sendEmailSummary(): Boolean
}

typealias Emailer = (message: String) -> Unit

interface Logger {
    fun info(message: String)
    fun error(message: String)
}

data class PipelineContext(val project: Project) {
    val hasTests: Boolean = project.hasTests()
    var testsRanSuccessfully = false
    var deployedSuccessfully = false
    fun mustRunDeployment(): Boolean {
        return testsRanSuccessfully || !hasTests
    }

    fun withTestsResults(testsResult: Boolean): PipelineContext {
        testsRanSuccessfully = testsResult
        return this
    }

    fun withDeploymentStatus(deploymentStatus: Boolean): PipelineContext {
        deployedSuccessfully = deploymentStatus
        return this
    }
}

class Project private constructor(private val buildsSuccessfully: Boolean, private val testStatus: TestStatus) {
    fun hasTests(): Boolean = testStatus != NoTests
    fun runTests(): String = if (testStatus == PassingTests) "success" else "failure"
    fun deploy(): String = if (buildsSuccessfully) "success" else "failure"

    class ProjectBuilder {
        private var buildsSuccessfully = false
        private var testStatus: TestStatus? = null
        fun setTestStatus(testStatus: TestStatus?): ProjectBuilder {
            this.testStatus = testStatus
            return this
        }

        fun setDeploysSuccessfully(buildsSuccessfully: Boolean): ProjectBuilder {
            this.buildsSuccessfully = buildsSuccessfully
            return this
        }

        fun build(): Project = Project(buildsSuccessfully, testStatus!!)
    }
}

enum class TestStatus {
    NoTests,
    PassingTests,
    FailingTests
}